src/engine.ml

```diff
@@ -611,15 +619,24 @@ let incoming_control_client config rng session channel now op data =
               let channel_st = Established keys in
               Ok (Some ip_config, config, { channel with channel_st }, tls_out)
           | None ->
-              (* now we send a PUSH_REQUEST\0 and see what happens *)
-              push_request tls' >>| fun (tls'', out) ->
-              let channel_st = Push_request_sent (tls'', key, tls_data) in
-              (* first send an ack for the received key data packet (this needs to be
-                 a separate packet from the PUSH_REQUEST for unknown reasons) *)
-              ( None,
-                config,
-                { channel with channel_st },
-                tls_out @ [ (`Ack, Cstruct.empty); (`Control, out) ] )))
+              let pull = Config.mem Pull config in
+              if pull then
+                let channel_st = Push_request_sent (tls', key, tls_data) in
+                Ok
+                  ( None,
+                    config,
+                    { channel with channel_st },
+                    tls_out @ [ (`Ack, Cstruct.empty) ] )
+              else
+                (* now we send a PUSH_REQUEST\0 and see what happens *)
+                push_request tls' >>| fun (tls'', out) ->
+                let channel_st = Push_request_sent (tls'', key, tls_data) in
+                (* first send an ack for the received key data packet (this needs to be
+                   a separate packet from the PUSH_REQUEST for unknown reasons) *)
+                ( None,
+                  config,
+                  { channel with channel_st },
+                  tls_out @ [ (`Ack, Cstruct.empty); (`Control, out) ] )))
```

Comment:
Do we actually want to send a push request if we don't `--pull`?

Reply:
This is, from what I saw, the current behavior of OpenVPN. `IV_PROTO=4` is activated only if `opt->pull` is set to `true`. It permits to keep the "old" behavior for us. Not sure if we want to keep it however.

Reply:
I haven't looked into PUSH_REQUEST and what it does, but from the name it sounds like something that is only useful with `--pull`. I may be very wrong of course :)

Reply:
indeed, in our state machine (and to behave well with a server), we need to send a push_request and get a push_reply to end up in established. we always do this in the "tls-client" mode (unless ifconfig is specified).

what is the reasoning behind it? to support a point-to-point connection, we don't need any push_request. everything is static (ip configuration, ...).

for a server that can deal with multiple clients, we need the client to have dynamically assigned an ip address (sent by the server, which knows which ips are already used). here a push_request is sensible.

below the line, there may be more configurational choice in terms of whether we need a push_request or not -- but it looks like the whole NCP_CIPHERS needs a push_reply in any case so the server can settle on a cipher to use for the data channel.